### DISCUSSION TITLE
Foundry Lottery: Funding Subscription on Local Chain

### DISCUSSION THREAD

**Discussion by spencerj411** [Cyfrin/foundry-full-course-cu#2290] [Q&A] [answered]
Hi, Cyfrin Updraft! A couple questions:

I'm currently on [Lesson 30 Fund Subscription](https://updraft.cyfrin.io/courses/foundry/smart-contract-lottery/fund-subscription) of Section 3: Smart Contract Lottery. Patrick is writing the following code:
```solidity
 function fundSubscription(address vrfCoordinatorV2_5, uint256 subId, address link, address account) public {
        if (block.chainid == LOCAL_CHAIN_ID) {
            vm.startBroadcast(account);
            VRFCoordinatorV2_5Mock(vrfCoordinatorV2_5).fundSubscription(subId, FUND_AMOUNT);
            vm.stopBroadcast();
        } else {
            vm.startBroadcast(account);
            LinkToken(link).transferAndCall(vrfCoordinatorV2_5, FUND_AMOUNT, abi.encode(subId));
            vm.stopBroadcast();
        }
    }
```

My question is, why do we need to use `VRFCoordinatorV2_5Mock().fundSubscription()` if we have already created a ` LinkToken()` mock? Can we not just run ` LinkToken(link).transferAndCall(vrfCoordinatorV2_5, FUND_AMOUNT, abi.encode(subId));` regardless of the `block.chainid`?

My second question is, how come we are calling the `LinkToken` contract **a mock** if it is needed for transferring LINK to the subscription even on mainnets and testnets?

Thanks in advance!

**Comment by EngrPips** [comment, marked as answer]
Hello @spencerj411, You have asked an interesting question. The main reason why we can't do `LinkToken(link).transferAndCall(vrfCoordinatorV2_5, FUND_AMOUNT, abi.encode(subId))` regardless of `block.chainid` is because, on sepolia chain, the `VRFCoordinatorV2` contract is an `ERC677Receiver` meaning we send an `ERC677` token to it and the `ERC677` token contract can at the same time call the the `onTokenTransfer` function on it. 

The below image shows the `VRFCoordinatorV2` contract, and I have put in a box the `onTokenTransfer` function that the `ERC677` token contract will call at the same time it transfers some token to it.

<img width="1440" alt="Screenshot 2024-08-22 at 10 13 26" src="https://github.com/user-attachments/assets/b822b072-57c5-47d9-8735-f691d6b0b905">

For our local Anvil Chain, our  `VRFCoordinatorV2Mock` contract is not an `ERC677Receiver` receiver, which means we can not just send token to it and have the token contract call the `onTokenTransfer` function as it does comply with that. Hence why we use the `fundSubscription` function to fund our subscription instead of sending token and having the token contract call the `onTokenTransfer` function on the `VRFCoordinatorV2Mock`

The image below shows the `VRFCoordinatorV2PlusMock` contract; you can clearly see that it does not have the `onTokenTransfer` function.

<img width="1440" alt="Screenshot 2024-08-22 at 10 34 14" src="https://github.com/user-attachments/assets/365d3de1-0908-4bad-a53f-df5ee17ac1ce">


**FOR YOUR SECOND TOKEN**
We were able to use the mock contract of the `Link` token to send token to the actual `Link` token contract on a real blockchain because we were only using the Interface of the mock contract to get the method we wanted to call on the address of a real `Link` contract deployed on a real blockchain.

`LinkToken(_linkToken).transferAndCall(_vrfCoordinatorContractAddress, TESTNET_FUND_AMOUNT, abi.encode(_subscriptionId));`

In the above code, we wrap the address of a real `Link` token contract deployed on some real blockchain with the interface of the mock `Link` token so that we can call the `transferAndCall` function on the real `Link` token contract.

If anything is confusing, Please ask questions.

**Reply by spencerj411** [reply]
@EngrPips you are awesome! Please correct me if any of my understanding is wrong:
**Question 1:** The `LinkToken ` transfers `ERC677` tokens (? the part I'm most unsure about). The VRF Coordinator (reference contract) on the Sepolia blockchain is a `ERC677Receiver `.  The `VRFCoordinatorV2_5Mock` mock reference contract we use when testing on the local chain is not a `ERC677Receiver `. Therefore, we wouldn't be able to use the same line of code to mock transfer LINK to our mock subscription.

**Question 2:** Essentially, the `LinkToken ` mock is repurposed (?) as an interface to the real VRF coordinator reference contract. Does that mean I can also just make an interface for the the VRF coordinator and replace it with the mock? Could you confirm the definition of mocks? My understanding was that it acts kind of like a stub to all reference contracts in our smart contract.

**Reply by EngrPips** [reply]
I think the second question was based on the `Link` token contract and not the `VRFCoordinatorV2_5Mock` contract. Basically, to form a contract object, you need an interface of the contract and the address of a deployed instance that abides by that interface. 

**EXAMPLE**`
```js
contractObject = Interface(address of a deployed instance)
```
So when we did the below, we were trying to form a contract object which we can then call a method on. In the below code, the `LinkToken` acts as the interface, and the `link` acts as the address of a deployed instance of a contract that abides by the `LinkToken` interface.  So the instruction was to use this `LinkToken` interface on this `link` contract address and call the `transferAndCall` function passing `vrfCoordinatorV2_5, FUND_AMOUNT, abi.encode(subId)` as the argument to the `transferAndCall` function.

```js
LinkToken(link).transferAndCall(vrfCoordinatorV2_5, FUND_AMOUNT, abi.encode(subId));
```
Just for clarity's sake, the `LinkToken` is an `ERC677` token. Aside clarifying this fact for the number 1 Question, you get the gist.

**Reply by spencerj411** [reply]
Apologies, that was a typo. When I said `VRFCoordinatorV2_5Mock` in question 2, I meant `LinkToken`. Apart from that, great! You explained it well. Thank you!

**Reply by EngrPips** [reply]
You are welcome, Friend. Keep crushing it.
